Switched to mutagen, third party keyboard support, added app icon, removed need for temporary files

PhilipTrauner opened the pull request:
Hey! Absolutely superb utility you've got there. Do you mind if I make it even better?

* Third party keyboard support  
    Third party keyboards sometimes emit different key-codes for their media-keys. I can't possibly test them all but this change should allow for next/previous support on all Logitech keyboards that use the Logitech Gaming Software.

* Switched to mutagen  
    All of my tracks have covers so I was a bit surprised when only a couple actually showed up. After some testing I threw out tinytag and switched to the more verbose but pretty "un-pythonic" mutagen. It works, I guess.

* Added app icon  
    The Python rocket is ugly. The new logo is at least a little better \\(^^)/

* Removed need for temporary files
    Only jpeg album art was supported by the temporary file solution. After some digging around in pyobjc I found a way to create NSImages dynamically in memory from binary data, instead of file-like objects.

I hope my changes don't break anything. Merge if you can find the time, please 😉.

PhilipTrauner committed:
Added third party keyboard keys

Necessary for Logitech keyboards, that apparently emit different key-codes for media

PhilipTrauner committed:
Added icon

Source: https://ryuutei.wordpress.com/2013/02/17/cmus-icon/
Slightly edited because the empty edges looked weird.

PhilipTrauner committed:
Install icon, replaced tinytag with mutagen

Some covers never showed up despite supposed format support by tinytag.
mutagen certainly has other flaws, but those are more convenient than the ones that tinytag has.

PhilipTrauner committed:
Updated install instructions to reflect the switch to mutagen

PhilipTrauner committed:
Removed information about icon_path

The icon_path is not needed anymore because the cover is converted to an NSImage in memory now.

PhilipTrauner committed:
No temporary files required anymore, switch to mutagen, added app icon

NSImages are now being generated in memory instead of creating temporary files.
PNG support for album artwork is basically free using this method.

Some covers never showed up despite supposed format support by tinytag.
mutagen certainly has other flaws, but those are more convenient than the ones that tinytag has.

The Python rocket is ugly. The new logo is at least a little better \(^^)/

azadkuh commented:
i really appreciate this, everything looks fine, except the `mutagen`. imho it's a good idea to make mutagen/tinytag an option at installation time.
please give me some time to review.

PhilipTrauner commented:
I've got an even bigger commit in the pipeline that gets rid of the bootstrapping process and instead utilises the cmus rc in addition to a python3 port. I'll add those changes to this pull request when I'm done, so you can just as well ignore it for now. I've gone a little overboard with the changes tough, and also removed most of the unneeded object orientation.

PhilipTrauner committed:
Added gitignore from @github/gitignore

PhilipTrauner committed:
Folder restructure.

PhilipTrauner committed:
Removed under-the-hood.md

The behaviour of cmus-osx has changed so drastically that under-the-hood.md became inaccurate.

PhilipTrauner committed:
Restructure, python3 rewrite, removed bootstrapping

All required files now reside in the subdirectory cmus-osx.
Installation is easier this way.

Refactored and rewrote most code and added python3 support in the process.
python2.7 is not supported anymore. Most object orientation was also ripped out, as it was unnecessary for one instance during runtime and slow.

Instead of starting cmus from a python script the cmus rc is utilised to start the media-key handler.

More changes to follow python coding conventions more closely.

PhilipTrauner committed:
Rewrote setup

Utilise the cmus rc for same-time startup, new directory structure.
cmus-osx now resides in ~/.config/cmus/cmus-osx

PhilipTrauner commented:
New commits added. I'm not sure if a merge is still the right way to go as I pretty much rewrote and refactored all code.

PhilipTrauner committed:
Fixed album seperator

PhilipTrauner committed:
Removed obsolete info, remade example gif
